### PR TITLE
Add 'setvar' and 'try_set_value' fuctions

### PR DESCRIPTION
**Pull request by dmitryilyin**
The can modify variables and data structures.

### CODE REVIEW

**Comment by nibalizer** [comment]
@dmitryilyin sorry, looks like we're in merge conflict. Can you update?

**Comment by nibalizer** [comment]
What is the correct form of modes for spec files?

File: README.markdown [hunk 1]
```diff
@@ -167,6 +167,67 @@ The result will be as follows:
...
+same random string for every host name.
+
+###setvar
+
+This function is the sibling of the "getvar" function. With its help you can
+actually *change* the value of a variable in the local scope. Yes, it have
+always been impossible to modify variables, but now you can.
+
+*Examples:*
+```ruby
```
Comment (nibalizer) [inline README.markdown]: You can use ```puppet actually

File: README.markdown [hunk 2]
```diff
@@ -167,6 +167,67 @@ The result will be as follows:
   notice: Scope(Class[main]): FtvfvkS9j9wXLsd6
 ```
 
+This functions is not "pure" and will return different value on every call.
```
Comment (nibalizer) [inline README.markdown]: If this is modifying the docs for the above function that should be a different commit.

File: README.markdown [hunk 3]
```diff
@@ -167,6 +167,67 @@ The result will be as follows:
...
+
+This function is limited only to variables local to the scope it's being used.
+If the variable you are modifying is being used as a reference by other classes
+it can lead to unexpected behaviour because the variable value will become parse
+order dependant, that should not usually happen nad does not work well at all with
+the declarative nature of the Puppet language. Yes, there are reasons you could
+not modify variables in the first place. So you can use this functions only when
+you know that it will not break anything.
+
+###try_set_value
```
Comment (nibalizer) [inline README.markdown]: What is the third argument to this?

**Comment by dmitryilyin** [comment]
Updated README with fixes

**Comment by dmitryilyin** [comment]
Hello?

**Comment by bastelfreak** [comment]
@dmitryilyin thanks for your contribution.

This branch has merge confilcts, could rebase it please?

File: README.markdown [hunk 4]
```diff
@@ -167,6 +167,71 @@ The result will be as follows:
...
+  'a' => {
+    'b' => [
+      'b1',
+      'b2',
+      'b3',
+    ]
+  }
+}
+
+$success = try_set_value($data, 'a/b/2', 'new_value', '/')
```
Comment (rnelson0) [inline README.markdown]: Is there a particular reason I'd want to use an array index here rather than the actual value I am trying to replace?

File: README.markdown [hunk 5]
```diff
@@ -167,6 +167,71 @@ The result will be as follows:
...
+order dependant, that should not usually happen nad does not work well at all with
+the declarative nature of the Puppet language. Yes, there are reasons you could
+not modify variables in the first place. So you can use this functions only when
+you know that it will not break anything.
+
+###try_set_value
+
+This function is the sibling of "try_get_value" function. Instead of retrieving
+the value from the complex structure it tries to set the value. It returns true
+if the value was set successfully and false if not.
```
Comment (rnelson0) [inline README.markdown]: Could an optional argument be provided so that compilation fails when false?

**Comment by rnelson0** [comment]
@dmitryilyin bump for comments and rebasing.

**Comment by ffrank** [comment]
Sorry for being the party pooper, but here goes.

It's kinda neat that this overcomes the limitation of variable immutability, because in some situations, it's just inconvenient.

However, I do believe that there has been careful deliberation in this language design decision. Off the top of my hat, I strongly feel that with these functions, we're treading in areas near `defined` and `ensure_resource`. Which is to say, they do add much convenience, at the cost of a grave danger of making manifest semantics evaluation order dependent. History taught us that this should be avoided whenever possible.

TL;DR I'm uncomfortable adding to the chain of dangerous tools.

**Comment by jyaworski** [comment]
What is the status of this?

**Comment by jyaworski** [comment]
Closing due to pushback and lack of responses. Please re-open if needed.
